File: src/main/java/fi/evident/dojo/raytracer/SceneObject.java
package fi.evident.dojo.raytracer;

public abstract class SceneObject {
    
    public final Surface surface;
    
    protected SceneObject(Surface surface) {
        assert surface != null;
        
        this.surface = surface;
    }
    
    /**
     * Returns the intersection of this object with given ray, or null
     * if the ray does not intersect the object.
     */
    public abstract Intersection intersect(Ray ray);
    
    /**
     * Returns the normal of the object at given position.
     */
    public abstract Vector3 normal(Vector3 pos);
}

File: src/main/kotlin/fi/evident/dojo/raytracer/SceneObject.kt
/*
 * Copyright (c) 2011-2012 Evident Solutions Oy
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in
 * all copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
 * THE SOFTWARE.
 */
package fi.evident.dojo.raytracer

import fi.evident.dojo.raytracer.math.Direction
import fi.evident.dojo.raytracer.math.Point

interface SceneObject {

    val surface: Surface

    /**
     * Returns the intersection of this object with given ray, or `null`
     * if the ray does not intersect the object.
     */
    fun intersect(ray: Ray): Intersection?

    /**
     * Returns the normal of the object at given position.
     */
    fun normal(pos: Point): Direction
}

